.github/workflows/retract.yml

```diff
@@ -6,7 +6,14 @@ jobs:
   retract:
     runs-on: ubuntu-latest
     name: Retract a release
-    if: ${{ !github.event.issue.pull_request && github.event.comment.body == '#retract' }}
+    # not a pull request and has `#contains` on a line by itself
+    # fromJSON is used to process escape sequences
+    if: |
+      !github.event.issue.pull_request &&
+      contains(
+        format(fromJSON('"\r\n{0}\r\n"'), github.event.comment.body),
+        fromJSON('"\r\n#retract\r\n"')
```

Comment:
I'm not a fan of this hack as it looks confusing. It also seems specific to Windows-style line endings?

Reply:
Why not just `contains(github.event.comment.body, '#retract')` and deal with it if not having it on a separate line becomes an issue?

Reply:
`<textarea>`s always have `\r\n` independent of platform

Reply:
Well, this is an API payload though not a direct read from a `<textarea>` and we certainly had issues around this earlier: https://github.com/getsentry/.github/pull/36

Reply:
ah I missed this comment entirely:

> Why not just `contains(github.event.comment.body, '#retract')` and deal with it if not having it on a separate line becomes an issue?

I have extensive experience building chatops systems and this becomes an issue almost immediately (either comments discussion the chatop or additional substrings afterwards) -- requiring it to be on its own line makes the operation explicit

Reply:
I don't think it will become an issue here as this is not a Slack or IRC channel and people rarely use `#something` randomly in their comments hence it is not worth having the extra complexity here but this is indeed not a blocker.

Reply:
TBH I'd just use

```
{{ !github.event.issue.pull_request && (startsWith(github.event.comment.body, '#retract') || endsWith(github.event.comment.body, '#retract')) }}
```

It doesn't add the complexity, yet it allows additional explanation before/after the retraction tag.

Reply:
TBH I'd just use

```
{{ !github.event.issue.pull_request && (startsWith(github.event.comment.body, '#retract') || endsWith(github.event.comment.body, '#retract')) }}
```

It doesn't add the complexity, yet it allows additional explanation before/after the retraction tag.

Reply:
That is true, but we are in full control here, so we can just point that out in the docs. And retracting is not a common use case to make the code harder to understand.

Tbh `format(fromJSON('"\r\n{0}\r\n"'), github.event.comment.body), fromJSON('"\r\n#retract\r\n"')` would take me some time to grasp without the comment.

Reply:
yeah totally agree that the code is hard to understand, but that's why the comment is there -- it's a rather unfortunate limitation in github actions.  I've got some inside contacts I'll bug about and see if they can't make this easier (even a `splitLines()` would go a long way here -- then it could just be `contains(splitLines(github.event.comment.body, '#retract')`)

Reply:
actually, the startsWith and endsWith doesn't work either because it doesn't check for a newline -- we would still need `fromJSON` to ensure there's a line afterwards

Reply:
>actually, the startsWith and endsWith doesn't work either because it doesn't check for a newline -- we would still need fromJSON to ensure there's a line afterwards

`contains()` it is then?

Reply:
contains is unsuitable as above because it can appear in sentences discussing retracting while not intending to actually retract

Reply:
@asottile-sentry 

https://github.com/getsentry/publish/pull/1095#discussion_r875133349

Would you like to use tail recursion optimization and either compromise or drop this PR as you don't seem to be convincing the people who needs to approve your PR?

Reply:
Gentlemen, let's keep it civil. @asottile-sentry is going to be supporting `publish` from now on, so I believe that this is his call by the end of the day.
Keep in mind that only a person who issued a release can retract it. And a chance of a single person typing backtick #retract backtick randomly in the comment is as big as someone putting a new line at the end/beginning of a message :P

Reply:
Hey, I enjoyed (mostly) craft/publish! :P

Reply:
>ah, @BYK -- I understand you're outside the picture so you don't have this context but my team is taking ownership of the craft/publish pipeline so @kamilogorek can focus on the projects they enjoy more

Well then, if _you_ are going to maintain this, who am I to judge 😅 Sorry for the earlier confusion and aggression. I just don't want to maintain (even remotely like this) some code I cannot justify for myself.

Reply:
>Hey, I enjoyed (mostly) craft/publish! :P

Sure sure :P